coretrace/simulation_data/cst_templates/heliostat.cpp

```diff
@@ -0,0 +1,437 @@
+                aim.set_values(0.0, 0.0, 2.0 * this->onaxis_canting_distance);
+            }
+            else if (this->canting_method == OFF_AXIS)
+            {
+                origin.set_values(panel_x, panel_y, 0.0);
+                // TODO: Set aim vector values
+                aim.set_values(0.0, 0.0, 1.0);
+                throw std::runtime_error("OFF_AXIS is not yet implemented");
+            }
+            else if (this->canting_method == UNSET)
+            {
+                throw std::runtime_error("Canting method unset");
+            }
+            else
+            {
+                throw std::runtime_error("Unknown canting method");
+            }
+
+            ap = make_aperture<Rectangle>(panel_len_x, panel_len_y);
+            elem->set_aperture(ap);
+
+            // // TODO: Is this the correct thing to do here?
+            // vector_add(1.0, this->focal_point, -1.0, origin, aim);
+
+            if (this->focal_length_x <= 0.0 && this->focal_length_y <= 0.0)
+            {
+                surf = make_surface<Flat>();
+                // Vector3d khat(0.0, 0.0, 1.0);
+                // vector_add(0.5, khat, 0.5, aim);
+            }
+            else
+            {
+                surf = make_surface<Parabola>(this->focal_length_x,
+                                              this->focal_length_y);
+            }
+            elem->set_surface(surf);
+
+            elem->set_reference_frame_geometry(origin,
+                                               aim,
+                                               0.0);
+
+            elem->set_front_optical_properties(this->optics_mirror);
+            elem->set_back_optical_properties(this->optics_mirror);
+            elem->enable();
+
+            this->heliostat_area += panel_len_x * panel_len_y;
+            this->facets.push_back(elem);
+            sts = this->add_element(elem);
+            if (!Element::is_success(sts))
+            {
+                throw std::runtime_error("Failed to add element");
+            }
+
+            panel_x += panel_len_x + this->gap_x;
+        }
+        panel_x = -0.5 * (this->aperture_size_x - panel_len_x);
+        panel_y += panel_len_y + this->gap_y;
+    }
+
+    this->initialized = true;
+
+    return;
+}
+
+void Heliostat::update_geometry(double azimuth,
+                                double elevation)
+{
+
+    if (elevation < 0.0 || elevation > 90.0)
+    {
+        std::stringstream ss;
+        ss << "Heliostat::update_geometry: Invalid elevation ("
+           << elevation << "). Elevation must lie between 0 and 90 degrees.";
+        throw std::invalid_argument(ss.str());
+    }
+
+    if (azimuth < -180.0 || azimuth > 180.0)
+    {
+        std::stringstream ss;
+        ss << "Heliostat::update_geometry: Invalid azimuth ("
+           << elevation << "). Azimuth must lie between -180 and 180 degrees.";
+        throw std::invalid_argument(ss.str());
+    }
+
+    this->coordinates_initialized = false;
```

Comment:
Should we add a check on this->initialized? that either calls create_geometry or throws an error.